SSA/Core/EDSLNested.lean

```diff
@@ -0,0 +1,257 @@
+import SSA.Core.WellTypedFramework
+/-
+an EDSL that elaborates expressions with nesting
+allowed into SSA.
+
+Authors: Siddharth Bhat
+-/
+namespace EDSL2
+open Lean Elab Macro
+
+declare_syntax_cat dsl_op2
+scoped syntax "[dsl_op2|" dsl_op2 "]" : term
+syntax dsl_var2 := "%v" num
+declare_syntax_cat dsl_expr2
+declare_syntax_cat dsl_region2
+declare_syntax_cat dsl_stmt2
+
+/-
+This is a fancier syntax, which allows expressions in the
+intermediate nodes. This is elaborated into SSA, with temporary
+variable names.
+-/
+syntax dsl_var2 : dsl_expr2
+scoped syntax "()" : dsl_expr2
+scoped syntax "(" dsl_expr2 ")" : dsl_expr2
+scoped syntax "(" dsl_expr2 "," dsl_expr2 ")" : dsl_expr2
+scoped syntax "(" dsl_expr2 "," dsl_expr2 "," dsl_expr2 ")" : dsl_expr2
+scoped syntax "op:" dsl_op2 ws dsl_expr2 (ws dsl_region2)? : dsl_expr2
+syntax dsl_assign2 := dsl_var2 ":= " dsl_expr2
+scoped syntax sepBy(dsl_assign2, ";") : dsl_stmt2
```

Comment:
a bit pedantic, but this is technically a collection of statements, no? and an assign a single statement?

Reply:
We call it `stmt` to mirror what we call the [`Index`](https://github.com/bollu/ssa/blob/main/SSA/Core/WellTypedFramework.lean#L165) in TSSA

```lean
inductive TSSAIndex (β : Type) : Type
/-- Ways of making an RHS -/
| EXPR : UserType β → TSSAIndex β
/-- A statement eventually returning a value of type `β` -/
| STMT : UserType β → TSSAIndex β
/-- a lambda -/
| REGION : UserType β → UserType β → TSSAIndex β
```